# Задача линейного программирования

## Вариант 2

### Исходные данные

$$
c = (-2, 1, 5, -3)
$$

$$
A =
\begin{pmatrix}
1 & -1 & 2 & 0 \\
0 & 3 & 1 & -2 \\
4 & 1 & 3 & 1
\end{pmatrix}
$$

$$
b = (8, 7, 15)
$$

Необходимо решить прямую задачу линейного программирования, построить двойственную задачу и проверить выполнение теоремы двойственности.

## 1. Прямая задача

Исходная задача имеет вид:

$$
F = -2x_1 + x_2 + 5x_3 - 3x_4 \rightarrow \max
$$

при ограничениях:

$$
\begin{cases}
x_1 - x_2 + 2x_3 \leq 8, \\
3x_2 + x_3 - 2x_4 \leq 7, \\
4x_1 + x_2 + 3x_3 + x_4 \leq 15, \\
x_1, x_2, x_3, x_4 \geq 0.
\end{cases}
$$

In [14]:
import numpy as np
from scipy.optimize import linprog

In [15]:
c = np.array([-2, 1, 5, -3], dtype=float)

A = np.array([
    [1, -1, 2,  0],
    [0,  3, 1, -2],
    [4,  1, 3,  1]
], dtype=float)

b = np.array([8, 7, 15], dtype=float)

print("c =", c)
print("A =")
print(A)
print("b =", b)

c = [-2.  1.  5. -3.]
A =
[[ 1. -1.  2.  0.]
 [ 0.  3.  1. -2.]
 [ 4.  1.  3.  1.]]
b = [ 8.  7. 15.]


In [16]:
# Решение прямой задачи

result_primal = linprog(
    c=-c,                  # linprog минимизирует, поэтому для max используем -c
    A_ub=A,
    b_ub=b,
    bounds=[(0, None)] * 4,
    method="highs"
)

print("Успешно решено:", result_primal.success)
print(
    "Оптимальное решение x =",
    [round(float(x), 6) for x in result_primal.x]
)
print(
    "Максимальное значение F =",
    round(float(-result_primal.fun), 6)
)

Успешно решено: True
Оптимальное решение x = [0.0, 0.857143, 4.428571, 0.0]
Максимальное значение F = 23.0


## 2. Двойственная задача

Для прямой задачи вида

$$
\max \; c^T x
$$

при условиях

$$
Ax \leq b, \qquad x \geq 0
$$

двойственная задача имеет вид

$$
\min \; b^T y
$$

при условиях

$$
A^T y \geq c, \qquad y \geq 0.
$$

Для варианта 2 двойственная задача имеет вид:

$$
G = 8y_1 + 7y_2 + 15y_3 \rightarrow \min
$$

при ограничениях:

$$
\begin{cases}
y_1 + 4y_3 \geq -2, \\
-y_1 + 3y_2 + y_3 \geq 1, \\
2y_1 + y_2 + 3y_3 \geq 5, \\
-2y_2 + y_3 \geq -3, \\
y_1, y_2, y_3 \geq 0.
\end{cases}
$$

In [17]:
# Решение двойственной задачи

result_dual = linprog(
    c=b,
    A_ub=-A.T,            # A^T y >= c  ->  -A^T y <= -c
    b_ub=-c,
    bounds=[(0, None)] * 3,
    method="highs"
)

print("Успешно решено:", result_dual.success)
print(
    "Оптимальное решение y =",
    [round(float(y), 6) for y in result_dual.x]
)
print(
    "Минимальное значение G =",
    round(float(result_dual.fun), 6)
)

Успешно решено: True
Оптимальное решение y = [2.0, 1.0, 0.0]
Минимальное значение G = 23.0


## 3. Проверка теоремы двойственности

Для прямой задачи было получено:

$$
x^* = (0,\;0.857143,\;4.428571,\;0)
$$

$$
F_{\max} = 23
$$

Для двойственной задачи:

$$
y^* = (2,\;1,\;0)
$$

$$
G_{\min} = 23
$$

Следовательно,

$$
F_{\max} = G_{\min} = 23.
$$

Таким образом, выполняется теорема двойственности линейного программирования.

## 4. Вывод

В работе была решена прямая задача линейного программирования и построена соответствующая двойственная задача.

Оптимальное решение прямой задачи:

$$
x^* = (0,\;0.857143,\;4.428571,\;0)
$$

при этом

$$
F_{\max}=23.
$$

Оптимальное решение двойственной задачи:

$$
y^*=(2,\;1,\;0)
$$

при этом

$$
G_{\min}=23.
$$

Так как

$$
F_{\max}=G_{\min}=23,
$$

теорема двойственности подтверждается.